### PR TITLE
Light Mode

### PR DESCRIPTION
**Pull request by malsf21**
This is a PR that implements a site-wide light mode and dark mode. Here's a summary of the PR:
* implements  a `themify()` SASS mixin and a `$themes` object to generate themed SASS classes (based on parent class specificity) (@emmyc)
* adds themed variants to the editor/output (@emmyc) and to the profile panel + sketches pages (@malsf21)
* creates a `<Switch>` custom component that adds visual decoration to a simple textbox (@emmyc, @fan-matt)
* adds the `<Switch>` component to the profile page to toggle the user's preference (@emmyc)
* slightly modify's the Profile Panel's name input (@malsf21)
* creates Redux handlers for the user's theme state (@fan-matt)
* saves user's state to document-wide cookies (@malsf21)
* writes Jest/Enzyme tests for the `<Switch>` component (@malsf21)

Big thanks to @emmyc for taking the lead on this PR, and @fan-matt for helping me correct my small mistakes!

### CODE REVIEW

**Comment by malsf21** [comment]
@emmyc when you can just take a look at the changes and let me know what you think!

@krashanoff @fan-matt @jamieliu386 if one of you can take a look when you're free to get some fresh eyes on this, that'd be great 😄 

👏  to @emmyc !!!!

File: src/styles/variables.scss [hunk 1]
```diff
@@ -13,9 +13,71 @@ $select-cyan: #83dadd;
...
+    .class{ 
+        @include themify($themes) {
+            background-color: themed("backgroundColor");
+        }
+    }
+*/
+
+$themes: (
+  light: (
+    backgroundColor: white !important,
```
Comment (malsf21) [inline src/styles/variables.scss]: In particular this is the color I think that should be different - I've tried `#ccc`, `#DDDCDF` and I think they offer more visibility but look less clean.

Comment (emmyc) [reply src/styles/variables.scss]: Yeah i completely agree! White was just easy for me to use as a placeholder, but definitely a little bit harsh? Same for the gutter font color, I think we could mute everything down a bit for the palette

Comment (malsf21) [reply src/styles/variables.scss]: Do you have any suggestions on which colors you think look the best! I feel like you've got a good eye for design 😮

**Comment by malsf21** [comment]
>Something I'm curious about, should we dispatch a theming action to the Redux store? If we were to get (a lot) more ambitious with our theming, it might come in handy. 

Hm, yeah. I think I'll end up hoisting the theme logic to the app-level, but it would make the most sense w/ a Redux action. I might make that another PR though, it could give someone a shot at handling a Redux ticket.

>  Somewhat related but not entirely, I think it'd do us a lot of good if we formalize our state organization.

Haha yes there's lots of work on that front. TBH, we don't use Redux as much as we should (mostly because I avoid using Redux unless I have to). I'm alright to take the lead on that, but if you are too then go for it (it'll be more fun for you than me).

**Comment by fan-matt** [comment]
> I'm alright to take the lead on that, but if you are too then go for it (it'll be more fun for you than me).

I'd be glad to help :)

>  I might make that another PR though, it could give someone a shot at handling a Redux ticket.

It's probably best to just go ahead and handle it now, there's a lot of opportunity for Redux PRs in the future. If you want, I can go ahead and make the changes

**Comment by malsf21** [comment]
If you'd like, totally good for you to add on to this PR. Everyone's getting in!

**Comment by malsf21** [comment]
Just hoisted the `theme-*` to the app-level, makes it easier to change the rest of the app's coloring. Will demo that now!

**Comment by malsf21** [comment]
Just committed a skeleton for light-mode themeing the output, with colors and stuff TBD.

**Comment by malsf21** [comment]
Oh also, we should write some tests 😄. Might come better w/ splitting the theme switcher into its own component.

**Comment by fan-matt** [comment]
> splitting the theme switcher into its own component.

If you're not doing that already, I can take care of that as well

**Comment by malsf21** [comment]
> > splitting the theme switcher into its own component.
> 
> If you're not doing that already, I can take care of that as well

Yeah you're good to do that too if you're interested. Looks like we'll wrap up this entire feature over the break and ship it before the quarter starts.

**Comment by fan-matt** [comment]
This update refactors the theme switch into a new component (Switch) that we can use in other places as well. Additionally, theme updates now happen through Redux

**Comment by malsf21** [comment]
Struggling still w/ finding the right colors for the light mode. Just updated the themeing to include the profile panel and the content pages; let me know what y'all think @krashanoff @fan-matt @emmyc.

**Comment by fan-matt** [comment]
> 
> 
> Struggling still w/ finding the right colors for the light mode. Just updated the themeing to include the profile panel and the content pages; let me know what y'all think @krashanoff @fan-matt @emmyc. Debating between a light lavender or the Teach LA green.

Green seems a bit off. What's the hex code for the lavender you're thinking of using?

**Comment by malsf21** [comment]
> > Struggling still w/ finding the right colors for the light mode. Just updated the themeing to include the profile panel and the content pages; let me know what y'all think @krashanoff @fan-matt @emmyc. Debating between a light lavender or the Teach LA green.
> 
> Green seems a bit off. What's the hex code for the lavender you're thinking of using?

Was considering #E6E7F9 but I have no idea tbh, open to any suggestions!

**Comment by fan-matt** [comment]
How's this one? #c0c0ce

Capture

**Comment by malsf21** [comment]
> #c0c0ce

Not bad, but tbh seems a bit too gray for me? Agh this is the worst part of creating frontends!

Also, other than the color scheme, how do y'all feel about the features implemented? Is there anything else that needs to be done? I'll write a few more tests for the Switch component.

**Comment by malsf21** [comment]
Pushed Matt's suggestion, other ones to try (which you can do by changing `accentColor` and `cardBorder` and `cardDivider` for the light theme (in `_variables.scss`)) are `$teachla-green` and `#E6E7F9`.

**Comment by fan-matt** [comment]
#9c9cce is a good one

**Comment by krashanoff** [comment]
Hey all, just digging into this and it all looks fantastic! Really impressed with all of this, nice work @emmyc! As for the remaining color scheming questions, I'm no designer, but I think that for the card borders and dividers at least, `$teachla-green` and `#E6E7F9` are both out - they are rather low contrast.

+1 on @fan-matt's  suggestion of `#9c9cce` for `accentBackground`. I was thinking along a similar vein with a deep salmon similar to the editor scheme: `#eb868d`, for example.

**Comment by fan-matt** [comment]
> #eb868d

ooooh lofi neapolitan

**Comment by malsf21** [comment]
Haha I also really like `#9c9cce`, I think it looks fresh! Do want to get @emmyc and @jamieliu386's opinions on this too, but I think the PR is good to merge other than the coloring. I know @emmyc mentioned something about making some of the whites less strong which I'm down for too.

**Comment by emmyc** [comment]
oop sorry i haven't checked this in a while, thank u all for fixing everything up!! she's truly a beaut :') 
i like the lavender! but im kinda feeling more of a deeper color to add contrast, what do yall think? (also i changed the white a smidge, if u can't tell that's a good sign)
image

**Comment by malsf21** [comment]
> oop sorry i haven't checked this in a while, thank u all for fixing everything up!! she's truly a beaut :')
> i like the lavender! but im kinda feeling more of a deeper color to add contrast, what do yall think? (also i changed the white a smidge, if u can't tell that's a good sign)
> image

Ooh yes actually I'm feeling the deeper colour too, I think it looks great! And I can only notice the difference in white because of GitHub's white directly beside it, but I agree that a slightly more muted white would be good. Do you mind pushing this? Should just involve changing `$light-lavender` and some of the background colours.

**Comment by fan-matt** [comment]
One thing to note before we merge, when you edit your name in the panel, the background is still the old panel color. If we want to change this, don't forget it!

**Comment by malsf21** [comment]
> One thing to note before we merge, when you edit your name in the panel, the background is still the old panel color. If we want to change this, don't forget it!

Right right, tbh I kinda dislike the behaviour of this box in general. I might push a fix to it tonight, but I'm also considering just making it transparent.

Also, it doesn't "save" your previous name - probably another issue to tackle another day.

**Comment by fan-matt** [comment]
Do we want color transitions like the sketches page has?

**Comment by malsf21** [comment]
> Do we want color transitions like the sketches page has?

I thought they looked jarring when applied to the editor, but I think it looks... better on sketches? Probably because there aren't a bunch of text tokens changing color. What do y'all think?

**Comment by malsf21** [comment]
Pushed a fix for the name editor in the Profile Panel, though there are tons of problems w/ the interface (but that's for another ticket).

**Comment by malsf21** [comment]
> Hey there. Just reviewed all the changes and, in my opinion, things look really solid! Frontend works as expected and the light mode switch and color scheme look great!
> 
> The only issue I found was when playing around with the viewports to make sure that the light mode switch was still accessible. I found it would get chopped off on viewports under ~580px in height:
> Screenshot_2019-12-15 Teach LA Editor
> That's scrolling all the way down. The same thing happens on the `master` build - it just has the advantage of having some more wiggle room with taller buttons. Of course the only feasible situation in which this would be an issue is on mobile devices, but I figured I should still raise this for consideration.

Good thing to point out. I think it's a tad out of scope for the PR, but I added it as a ticket in our Trello.

**Pull request merged**